redis/runtime/src/main/java/io/quarkiverse/langchain4j/redis/RedisEmbeddingStore.java

```diff
@@ -0,0 +1,246 @@
+                    .arg(schema.getPrefix())
+                    .arg("SCHEMA");
+            schema.defineFields(request);
+            LOG.debug(
+                    "Creating index with command: " + request.toString().replaceAll("\r\n", " "));
+            ds.getRedis().send(request).await().indefinitely();
+        } else {
+            LOG.debug("Index in Redis already exists: " + schema.getIndexName());
+        }
+    }
+
+    @Override
+    public String add(Embedding embedding) {
+        String id = randomUUID();
+        add(id, embedding);
+        return id;
+    }
+
+    @Override
+    public void add(String id, Embedding embedding) {
+        addInternal(id, embedding, null);
+    }
+
+    @Override
+    public String add(Embedding embedding, TextSegment textSegment) {
+        String id = randomUUID();
+        addInternal(id, embedding, textSegment);
+        return id;
+    }
+
+    @Override
+    public List<String> addAll(List<Embedding> embeddings) {
+        List<String> ids = embeddings.stream()
+                .map(ignored -> randomUUID())
+                .collect(toList());
+        addAllInternal(ids, embeddings, null);
+        return ids;
+    }
+
+    @Override
+    public List<String> addAll(List<Embedding> embeddings, List<TextSegment> embedded) {
+        List<String> ids = embeddings.stream()
+                .map(ignored -> randomUUID())
+                .collect(toList());
+        addAllInternal(ids, embeddings, embedded);
+        return ids;
+    }
+
+    private void addInternal(String id, Embedding embedding, TextSegment embedded) {
+        addAllInternal(singletonList(id), singletonList(embedding), embedded == null ? null : singletonList(embedded));
+    }
+
+    private void addAllInternal(List<String> ids, List<Embedding> embeddings, List<TextSegment> embedded) {
+        if (ids.isEmpty() || ids.size() != embeddings.size() || (embedded != null && embedded.size() != embeddings.size())) {
+            throw new IllegalArgumentException("ids, embeddings and embedded must be non-empty and of the same size");
+        }
+        ds.withTransaction(ds -> {
+            ReactiveTransactionalJsonCommands<String> json = ds.json();
+            int size = ids.size();
+            Uni[] unis = new Uni[size];
+            for (int i = 0; i < size; i++) {
+                String id = ids.get(i);
+                Embedding embedding = embeddings.get(i);
+                TextSegment textSegment = embedded == null ? null : embedded.get(i);
+                Map<String, Object> fields = new HashMap<>();
+                fields.put(schema.getVectorFieldName(), embedding.vector());
+                if (textSegment != null) {
+                    fields.put(schema.getScalarFieldName(), textSegment.text());
+                    fields.putAll(textSegment.metadata().asMap());
+                }
+                String key = schema.getPrefix() + id;
+                unis[i] = json.jsonSet(key, "$", fields);
```

Comment:
If isolation is important, I would use the transaction but with a watch on `key`. So, if the key is modified, the transaction is rolled back (well, discarded as there is no rollback).